vivisect/extensions/__init__.py

```diff
@@ -14,33 +15,83 @@
 (vw) and an vivisect gui reference (if present).
 '''
 
-def loadExtensions( vw, vwgui ):
+logger = logging.getLogger(__name__)
 
+
+def earlyExtensions(vw):
+    logger.debug('earlyLoad')
+    for mod in extensions:
+        if not hasattr(mod, 'earlyLoad'):
+            logger.debug('earlyLoad: %r (no earlyLoad() function)', mod)
+            continue
+
+        logger.debug('earlyLoad: %r', mod)
+        try:
+            mod.earlyLoad(vw)
+        except Exception, e:
+            vw.vprint(traceback.format_exc())
+            vw.vprint('Extension Error (early): %s' % filepath)
+
+
+def preFileLoadExtensions(vw, fname, bytez, fd):
+    logger.debug('preFileLoad')
+    for mod in extensions:
+        if not hasattr(mod, 'preFileLoadHook'):
+            logger.debug('preFileLoad: %r (no preFileLoadHook() function)', mod)
+            continue
+
+        logger.debug('preFileLoad: %r', mod)
+        try:
+            mod.preFileLoadHook(vw, fname, bytez, fd)
+        except Exception, e:
+            vw.vprint(traceback.format_exc())
+            vw.vprint('Extension Error (preFileLoad): %s' % filepath)
+
+
+def loadExtensions(vw, vwgui):
+    logger.debug('loadExtensions')
+    for mod in extensions:
+        logger.debug('loadExtensionsL %r', mod)
+        try:
+            mod.vivExtension(vw, vwgui)
+        except Exception, e:
+            vw.vprint(traceback.format_exc())
+            vw.vprint('Extension Error: %s' % filepath)
+
+extensions = []
+def impExtensions():
+    global extensions
     extdir = os.getenv('VIV_EXT_PATH')
 
-    if extdir == None:
+    if extdir is None:
         return
 
+    logger.info('impExtensions: VIV_EXT_PATH == %r', extdir)
     for dirname in extdir.split(';'):
+        logger.info('impExtensions: %r', dirname)
 
         if not os.path.isdir(dirname):
-            vw.vprint('Invalid VIV_EXT_PATH dir: %s' % dirname)
+            logger.warn('Invalid VIV_EXT_PATH dir: %s' % dirname)
             continue
 
         for fname in os.listdir(dirname):
 
             if not fname.endswith('.py'):
                 continue
 
+            logger.info('impExtensions: %s/%s', dirname, fname)
             # Build code objects from the module files
             mod = imp.new_module('viv_ext')
             filepath = os.path.join(dirname, fname)
-            filebytes = file( filepath, 'r' ).read()
+            filebytes = file(filepath, 'r').read()
             mod.__file__ = filepath
             try:
                 exec filebytes in mod.__dict__
-                mod.vivExtension(vw, vwgui)
-            except Exception, e:
-                vw.vprint( traceback.format_exc() )
-                vw.vprint('Extension Error: %s' % filepath)
+                extensions.append(mod)
+                
+            except Exception as e:
+                logger.warn(traceback.format_exc())
```

Comment:
you can use `logging.somelevel("....", args, exc_info=True)` to render the traceback with the log message

Reply:
also, looks like inconsistent reporting: is it `vw.vprint` or `logger.warn` on extension errors?